File: python/2021/15_chiton/aoc202115.py
"""AoC 15, 2021: Chiton"""

# Standard library imports
import heapq
import itertools
import pathlib
import sys

# Third party imports
import colorama
import numpy as np
from colorama import Cursor, Fore

colorama.init()


def parse(puzzle_input):
    """Parse input"""
    return np.array(
        [[int(energy) for energy in line] for line in puzzle_input.split("\n")]
    )


def part1(data):
    """Solve part 1"""
    return find_cheapest(data)


def part2(data):
    """Solve part 2"""
    return find_cheapest(expand(data, 5, 5))


def expand(risks, num_tile_rows, num_tile_cols):
    """Expand risk map by tiling risks

    >>> expand(np.array([[1, 4], [8, 9]]), 2, 2)
    array([[1, 4, 2, 5],
           [8, 9, 9, 1],
           [2, 5, 3, 6],
           [9, 1, 1, 2]])
    """
    num_rows, num_cols = risks.shape
    tiled = np.tile(risks, (num_tile_rows, num_tile_cols))
    for row, col in itertools.product(range(num_tile_rows), range(num_tile_cols)):
        tiled[
            num_rows * row : num_rows * (row + 1), num_cols * col : num_cols * (col + 1)
        ] += (row + col)

    tiled[tiled > 9] -= 9
    return tiled


def find_cheapest(risks):
    """Find the cost of the cheapest path through the risk levels

    Use a heap to keep track of the cheapest next step.

    [1]:2: 9          --[2]:9:         |  X :9:         |  X :9:
               2,3             3,9,11  |        8,9,11  |
    :3: 7  1   -->   :3::7: 1   -->   [3]:7: 1   -->    | :7: 1
                                                        |
     5  8  3          5  8  3         :5: 8  3         [5]:8: 3


              ---+ :9:          ---+ :9:          -----[9]         ---+  X
     9,11,16     |     10,11,16    |     11,13,16           13,16     |
       -->    X [7]:1:   -->    X  +-[1]   -->    X  X  X    -->   X  +--+   -->   13
                                                                         |
              X :8: 3           X :8::3:          X :8::3:         X :8:[3]

    >>> find_cheapest(np.array([[1, 2, 9], [3, 7, 1], [5, 8, 3]]))
    13
    """
    num_rows, num_cols = risks.shape
    target = (num_rows - 1, num_cols - 1)
    locations = [(0, (0, 0))]
    seen = set()

    while True:
        cost, current = heapq.heappop(locations)
        if "-v" in sys.argv:
            visualize(risks, seen, current)

        if current == target:
            return cost

        for neighbor in neighbors(current, *risks.shape):
            if neighbor not in seen:
                seen.add(neighbor)
                heapq.heappush(locations, (cost + risks[neighbor], neighbor))


def neighbors(current, num_rows, num_cols):
    """Find neighbors to the current position

    >>> list(neighbors((2, 2), 5, 5))
    [(1, 2), (2, 1), (2, 3), (3, 2)]

    >>> list(neighbors((0, 4), 5, 5))
    [(0, 3), (1, 4)]
    """
    row, col = current

    if row - 1 >= 0:
        yield (row - 1, col)
    if col - 1 >= 0:
        yield (row, col - 1)
    if col + 1 < num_cols:
        yield (row, col + 1)
    if row + 1 < num_rows:
        yield (row + 1, col)


def visualize(risks, seen, current):
    """Draw risks to terminal"""
    for row, line in enumerate(risks):
        for col, risk in enumerate(line):
            symbol = (
                Fore.RED + "#"
                if (row, col) == current
                else Fore.YELLOW + str(risks[row, col])
                if (row, col) in seen
                else Fore.GREEN + str(risks[row, col])
            )
            print(Cursor.POS(col + 1, row + 1) + symbol)


def solve(puzzle_input):
    """Solve the puzzle for the given input"""
    data = parse(puzzle_input)
    solution1 = part1(data)
    solution2 = part2(data)

    return solution1, solution2


if __name__ == "__main__":
    for path in sys.argv[1:]:
        if path.startswith("-"):
            continue

        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))

File: elixir/lib/2021/15_chiton/aoc202115.ex
defmodule AOC2021.Day15 do
  @moduledoc """
  Advent of Code 2021, day 15: Chiton

  Appreciation and apologies to: https://github.com/josevalim/aoc/blob/main/2021/day-15.livemd
  """
  require AOC

  @doc """
  Parse input
  """
  def parse(puzzle_input) do
    for {line, row} <- Enum.with_index(String.split(puzzle_input, "\n", trim: true)),
        {char, col} <- Enum.with_index(String.split(line, "", trim: true)),
        risk = String.to_integer(char),
        into: %{},
        do: {{row, col}, risk}
  end

  @doc """
  Solve part 1
  """
  def part1(risks), do: risks |> find_safest()

  @doc """
  Solve part 2
  """
  def part2(risks), do: risks |> expand({5, 5}) |> find_safest()

  @doc """
  Use Dijkstra to find the total risk of the safest path through the grid.

  ## Example:

      321      321
      456  ->  ..6
      987      ..7

      iex> risks = parse("321\\n456\\n987")
      iex> find_safest(risks)
      16
  """
  def find_safest(risks) do
    total_risks = %{{0, 0} => 0}
    queue = Heap.min() |> Heap.push({0, {0, 0}})
    target = risks |> Map.keys() |> Enum.max()

    dijkstra(risks, total_risks, queue, target)
  end

  defp dijkstra(risks, total_risks, queue, target) do
    {{risk, {row, col} = node}, queue} = Heap.split(queue)

    if node == target do
      risk
    else
      neighbors = [{row - 1, col}, {row + 1, col}, {row, col - 1}, {row, col + 1}]

      {queue, total_risks} =
        for next <- neighbors,
            Map.has_key?(risks, next),
            next_risk = risk + risks[next],
            next_risk < Map.get(total_risks, next, :infinity),
            reduce: {queue, total_risks} do
          {queue, total_risks} ->
            queue = Heap.push(queue, {next_risk, next})
            total_risks = Map.put(total_risks, next, next_risk)
            {queue, total_risks}
        end

      dijkstra(risks, total_risks, queue, target)
    end
  end

  @doc """
  Expand a grid while increasing the risk level.

  ## Example:
                12|23|34
      12        89|91|12
      89   ->   --+--+--
                23|34|45
                91|12|23

      iex> expand(%{{0, 0} => 1, {0, 1} => 2, {1, 0} => 8, {1, 1} => 9}, {2, 3})
      %{{0, 0} => 1, {0, 1} => 2, {1, 0} => 8, {1, 1} => 9,
        {2, 0} => 2, {2, 1} => 3, {3, 0} => 9, {3, 1} => 1,
        {0, 2} => 2, {0, 3} => 3, {1, 2} => 9, {1, 3} => 1,
        {2, 2} => 3, {2, 3} => 4, {3, 2} => 1, {3, 3} => 2,
        {0, 4} => 3, {0, 5} => 4, {1, 4} => 1, {1, 5} => 2,
        {2, 4} => 4, {2, 5} => 5, {3, 4} => 2, {3, 5} => 3}
  """
  def expand(risks, {height, width}) do
    {max_row, max_col} = risks |> Map.keys() |> Enum.max()
    {rows, cols} = {max_row + 1, max_col + 1}

    for x <- 0..(width - 1), y <- 0..(height - 1), reduce: risks do
      acc ->
        for row <- 0..max_row, col <- 0..max_col, reduce: acc do
          acc ->
            Map.put(acc, {row + rows * y, col + cols * x}, bump(risks[{row, col}], x + y))
        end
    end
  end

  @doc """
  Bump the risk level.

  ## Example:

      iex> bump(3, 4)
      7
      iex> bump(5, 4)
      9
      iex> bump(7, 3)
      1
  """
  def bump(value, add), do: rem(value + add - 1, 9) + 1

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end
